kaminariを使用したページング処理の提出物

ham-cap opened the pull request:
ページング処理を実装しました。
ご確認のほどよろしくお願いいたします。

ham-cap committed:
kaminariを使用してページング処理を実装した・ページングの表記を日本語化した

JunichiIto commented on config/application.rb:
```diff
@@ -18,5 +18,7 @@ class Application < Rails::Application
     #
     # config.time_zone = "Central Time (US & Canada)"
     # config.eager_load_paths << Rails.root.join("extras")
+    config.i18n.load_path += Dir[Rails.root.join('config', 'locales', '*.{rb, yml}')]
```
質問）この行は何のために追加されました？これがないと何か問題がありましたか？
もし参考にしたwebページ等があれば教えてください〜。

ham-cap replied:
プラクティスの歓迎要件にある
> kaminariの"First"や"Last"のような表示を日本語化する

に取り組む際、i18nを使用してロケールを参照するために必要という認識だったため追加しました。
この行と直下のデフォルトロケールの設定はワンセットと思っていたため、この行が無い場合というのは検証しておりません。
この行がi18nに対して`config/locales`にある`*.rb`あるいは`*.yml`ファイルを参照するよう指示して、次の行でデフォルトの参照ファイルを指定しているという認識です。

参考にしたページはRailsガイドの以下のページです。
https://railsguides.jp/i18n.html#rails%E3%82%A2%E3%83%97%E3%83%AA%E3%82%B1%E3%83%BC%E3%82%B7%E3%83%A7%E3%83%B3%E3%82%92%E5%9B%BD%E9%9A%9B%E5%8C%96%E5%90%91%E3%81%91%E3%81%AB%E8%A8%AD%E5%AE%9A%E3%81%99%E3%82%8B

上記の認識に誤りがあればご教示いただければ幸いです。

JunichiIto replied:
> 参考にしたページはRailsガイドの以下のページです。

なるほど〜。では、このセクションをもう一度最初からよく直してみると・・・？？

> この行と直下のデフォルトロケールの設定はワンセットと思っていたため、この行が無い場合というのは検証しておりません。

~~そんなに大変な検証ではないと思うので、「検証しておりません」というコメントを返す前に一度試してみてほしかったな〜と思いました。「ワンセットと思っていた」という点もRailsガイドにをよく読むと決してワンセットで設定しろ、とは書いてないです。~~
~~文章よりもサンプルコードに先に食いついてしまう傾向があるように思えるので（まあ、僕もたまにやっちゃいますが😅）、Railsガイドの本文もよく読んでみましょう〜。~~

（すぐに検証できない状況だったとのことなので）この行はなくても動くはずなので、確認の上、コードを修正してください。

ham-cap replied:
もう一度アタマから読んでみました。
> config/locales以下にあるすべての.rbファイルと.ymlファイルは、自動的に訳文読み込みパスに追加されます。

とあったので、ご指摘いただいた行をわざわざ書く必要はないのですね。失礼いたしました。
この部分は何らかの理由で読み込みパスに`config/locales`以外のディレクトリ内にあるファイルを読み込ませたい場合に使うものと理解しました。
仰るとおり特にワンセットとも書かれていないようですし、今後はもう少しよく読むようにします。

> そんなに大変な検証ではないと思うので、「検証しておりません」というコメントを返す前に一度試してみてほしかったな〜と思いました。

少し引っかかったので一点だけ。
出先からコメントを書き込んだためその場での検証はできていませんでしたが、後でちゃんとしています。
たしかに検証してから返信しても何ら問題はなかったと思いますので、そこが前後してしまったのは反省していますが、ご質問は「これがないと何か問題がありましたか？」という過去形でしたので、あくまでも今回のPRを投げるまでの実装段階でそういった検証をしたかどうか、していた場合の結果はどうだったか、を確認されていると理解しました。
そのため、（実装した段階では）検証していませんという回答になったまでです。
私自身はこの程度の検証の手間は惜しみませんし、短絡的にただ回答だけをもらおうとも思っていません。
しかし上記のような書き方をされると、仮に第三者がこのPRを読んだ場合私がたった一行の検証の手間を惜しみ、自分で考えずに回答だけもらおうとしているかのような印象を与えてしまう可能性が高いのではないでしょうか？

ham-cap replied:
Railsガイドをもう一度アタマから読んでみました。

> config/locales以下にあるすべての.rbファイルと.ymlファイルは、自動的に訳文読み込みパスに追加されます。

とあったので、ご指摘いただいた行をわざわざ書いて`config/locales`を指定する必要はないのですね。
失礼いたしました。
自分なりに考えてみましたが、この部分は何らかの理由で読み込みパスに`config/locales`以外のディレクトリ内にあるファイルを読み込ませたい場合に使うものと理解しました。（いまのところ具体例までは思い浮かびませんのでまた調べてみようかと思います。）
また、仰るとおり直下のデフォルトロケールを指定する行とワンセットとも書かれていないようですし、この点は完全に私の早とちりによる思い込みでした。今後はもう少しよく読むようにします。

> そんなに大変な検証ではないと思うので、「検証しておりません」というコメントを返す前に一度試してみてほしかったな〜と思いました。

少し引っかかったので一点だけ。
出先から返信したためその場での検証はできていませんでしたが、後でちゃんとしています。
たしかに検証してから返信しても問題はなかったと思いますので、そこが前後してしまったのは反省していますが、ご質問は「これがないと何か問題がありましたか？」という過去形でしたので、あくまでも今回のPRを投げるまでの実装段階でそういった検証をしたかどうか、していた場合の結果はどうだったか、を確認されていると理解しました。そのため、（実装した段階では）検証していませんという意図で回答しました。
この部分は私の返信もかなり言葉足らずで申し訳ありませんでした。
平日の日中は仕事をしており、自分のPCが目の前にあるわけではないのですぐに検証はできませんが、休憩中にGitHubを覗くぐらいはできますので、その時点で回答できるものは基本的に保留にせずに回答するようにしています。
仮に「これがない場合、何か問題がありますか？」と現在形で問われれば、その時点では回答材料が自分の中にありませんので、検証後に返信するという対応になると思います。

私自身はこの程度の検証の手間は惜しみませんし、短絡的にただ回答だけをもらおうとも思っていません。
自分が好きでしていることとはいえ、毎日時間を削り出しながら学習していますので、たった一行の検証の手間すら惜しみ、自分で考えずに回答だけもらおうとしているような印象になってしまっているとすれば大変不本意ですし残念です。
とはいえ、個人的にそう思われているだけなら特に実害はないので構わないのですが、このPRのやりとりは後で第三者の目にふれる可能性があり、その際に上記の点について私から何も説明がなされていなければ、私の課題への取り組み姿勢そのものに疑問符がついてしまいかねず、そういった事態は看過できませんので一応この場で説明だけはさせていただきました。

以上、よろしくお願いいたします。

JunichiIto replied:
うーむ、なるほど、これはオンラインコミュニケーションの難しさを表す事例ですねえ。

前回のやりとりでは、ham-capさんのコンテキスト（置かれている立場や状況）がうまく共有できていなかったところにすれ違いの一番の原因があるように思いました（だからham-capさんが悪い、という話に持っていこうとしてるのではなく、これはあくまで原因の分析です）。

そのコンテキストとは、具体的にいうとこの部分です。

> 出先から返信したためその場での検証はできていませんでした

この状況を僕が全く把握できておらず、今初めて知ったんですよね。

メンターをやってると日中でもふつうに学習を進めている生徒さんもたくさんいるので、他の生徒さんと同様にham-capさんも「質問の返事を返してくる＝PCの前にいてすぐにでもコードが触れる状態にある」と考えていました。

それゆえ「コメントを返す前に一度試してみてほしかったな〜と思いました」と発言した次第です。

加えて、僕の「これがないと何か問題がありましたか？」というのは、「これがないと何か問題がありましたか？ **（もし確認してないなら確認してみてくださいね）** 」という含みを持たせていたものだったので、ここもちょっとハイコンテキストだったかもしれません。過去形か現在形かでニュアンスはたしかに変わりそうですが、僕の中ではそれよりも「もし確認してないなら確認してみてくださいね」に重きを置いていました（が、そこを明文化していないのでこの文章もハイコンテキストです）。

で、僕としては「もし確認してないなら確認してみてくださいね」を期待していたのに、「この行が無い場合というのは検証しておりません」と返されたので、僕から見ると、「お、おおう・・・」と思ってしまいました。

が、これはどっちが悪いと言うよりも、お互いにコンテキストを上手に共有できていなかったゆえの悲劇（？）ではないかなと思います。
こういった問題をうまく解決するのは難しいですが、オンラインでやりとりするときは相手に見えていなさそうな自分の状況や考えを積極的に明文化する、ということを心がけるしかないのかなーと思います。

と、僕の中ではこんな着地点を考えてみたのですが、ham-capさんの方はいかがでしょうか？まだしっくりきていない部分があれば遠慮無くおっしゃってください。

ham-cap replied:
> これはどっちが悪いと言うよりも、お互いにコンテキストを上手に共有できていなかったゆえの悲劇（？）ではないかなと思います。

私もそう思います。
返信する際に一言、PCの前にいるわけではないということを説明するか、実装の段階では検証していませんが後でやってみます、のような感じで伝えておけばよかったかなと思っています。

> メンターをやってると日中でもふつうに学習を進めている生徒さんもたくさんいるので、他の生徒さんと同様にham-capさんも「質問の返事を返してくる＝PCの前にいてすぐにでもコードが触れる状態にある」と考えていました。

おそらくこう思われているのだろうなというのは最初の返信をいただいたときに分かりました。
自分が逆の立場でも「質問の返事を返してくる＝PCの前にいてすぐにでもコードが触れる状態にある」とあん

ham-cap replied:
> これはどっちが悪いと言うよりも、お互いにコンテキストを上手に共有できていなかったゆえの悲劇（？）ではないかなと思います。

基本的には私もそう思います。
最初に返信する際に一言、PCの前にいるわけではないということを説明するなり、実装の段階では検証していませんが後でやってみます、のような感じで伝えておくなりすればよかったかなと反省しています。

> メンターをやってると日中でもふつうに学習を進めている生徒さんもたくさんいるので、他の生徒さんと同様にham-capさんも「質問の返事を返してくる＝PCの前にいてすぐにでもコードが触れる状態にある」と考えていました。

おそらくこう思われているのだろうなというのは最初の返信をいただいたときに分かりました。
もちろん、多くの場合そのパターンに当てはまりそうだということも理解できますし、自分が逆の立場でも「質問の返事を返してくる＝PCの前にいてすぐにでもコードが触れる状態にある」と暗黙的に理解してしまう可能性はあるので、仕方ない部分もあるとは思いますが、同時に他の可能性を安易に排除し過ぎだとも思います。

私が今回一番引っかかっているのはこのあたりです。
様々な環境の生徒がいるという認識はあるはずなのに、「質問の返事を返してくる」という行動に対応する状況が「PCの前にいてすぐにでもコードが触れる状態にある」の1パターンしかないと考えるのは無理があります。
今回の私のように、返信はできるけどまだコードは触れないということはさほどレアケースとも思えませんし、他にも人によって状況は様々ですので。
にもかかわらず、他の可能性を勘案せずに「コメントを返す前に一度試してみてほしかったな〜と思いました」という書き方をされてしまうと、何か先入観を持たれているように感じてしまい、私でなくてもいい気持ちはしないのではないかなと思います。
上の返信でも書きましたが、ほんの少しの検証の手間すら惜しむ思考停止の生徒と認識されているように感じるからです。これは私の人間性に関わる部分ですのでなおさらです。
また、これも先の返信と重複しますが、第三者から見てもそういうネガティブな印象にとられやすい書き方を、第三者が見られる空間でされるのはもう少し慎重にしていただきたいと思います。
少なくとも、今回はコミュニケーションの行き違いで生じた誤解であるという理解で一致するのであれば、明確に訂正していただきたい点です。

> こういった問題をうまく解決するのは難しいですが、オンラインでやりとりするときは相手に見えていなさそうな自分の状況や考えを積極的に明文化する、ということを心がけるしかないのかなーと思います。

この点についても同意します。
そもそもテキストベースのコミュニケーションや業務上のやり取りにおいては明文化せずに「暗黙的に理解してしまう」ことや「ハイコンテキストである」ことは避けるべきことだと実感しました。

もう少し詳細に自分なりに整理してみると、
情報を伝える側は、
- 自分の状況をできるだけ相手に共有する
- 相手にしてほしいことは明確に書いて伝える（コンテキストや含みに頼りすぎない）

という点、
情報を受け取る側は、

- 相手の状況等を含め、書かれていないことを脳内で補完して断定しない

という点にそれぞれ注意する。
といったところかなと思うのですが、いかがでしょう。

全体的には、こうして文字だけでやり取りをする際は少し丁寧すぎるぐらいで丁度いいのだろうなと改めて感じました。（何を言いたいかわからなくなるリスクもありますが。）

JunichiIto replied:
たしかに、いろんな状況の生徒さんがおられますよね。

（プライベートな情報なので細かくは書きませんが）ham-capさんのプロフィールを読んで、忙しい中、隙間時間を見つけて工夫して勉強されてるんだなあということがよくわかりました。
仕事中の休憩時間も使ってがんばって勉強されている生徒さんがいることをしっかりと覚えておきますね！

> といったところかなと思うのですが、いかがでしょう。
全体的には、こうして文字だけでやり取りをする際は少し丁寧すぎるぐらいで丁度いいのだろうなと改めて感じました。

そうですね。おっしゃるとおりだと思います。

件のコメントについてはいったん取り消し線を引いた上で、「すぐに検証できない状況だとわかっていたらどうコメントするか？」というのを考えて書き直してみました。

こんな感じでいかがでしょうか？まるまる削除してほしい、等、何かリクエストがあればおっしゃってください。

ham-cap replied:
ありがとうございます。
取り消し線を引いていただいた状態で差し支えありません。
今回の一連のやりとりは大変勉強になりました。
面倒臭がらずに真摯にご対応いただいて感謝しています。

後ほど修正のコミットをプッシュしますので、改めてご確認のほどよろしくお願いいたします。

JunichiIto reviewed (commented):
1点質問しました〜。

ham-cap committed:
訳文読み込みパスの設定部分を削除

訳文読み込みパスにconfig/localesを設定していたが、config/localesは何も設定しなくても
自動的に読み込みパスに追加されており、あえて設定する必要がないため削除した。

JunichiIto reviewed (approved):
確認しました。OKです〜。
